hazelcast/src/main/java/com/hazelcast/nio/tcp/BindHandler.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.nio.tcp;
+
+import com.hazelcast.instance.ProtocolType;
+import com.hazelcast.internal.cluster.impl.BindMessage;
+import com.hazelcast.internal.cluster.impl.ExtendedBindMessage;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.nio.Address;
+import com.hazelcast.nio.Connection;
+import com.hazelcast.nio.IOService;
+import com.hazelcast.nio.Packet;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+final class BindHandler {
+
+    private final TcpIpEndpointManager tcpIpEndpointManager;
+    private final IOService ioService;
+    private final ILogger logger;
+    private final boolean spoofingChecks;
+
+    private final Set<ProtocolType> supportedProtocolTypes;
+
+    BindHandler(TcpIpEndpointManager tcpIpEndpointManager, IOService ioService, ILogger logger,
+                boolean spoofingChecks, Set<ProtocolType> supportedProtocolTypes) {
+        this.tcpIpEndpointManager = tcpIpEndpointManager;
+        this.ioService = ioService;
+        this.logger = logger;
+        this.spoofingChecks = spoofingChecks;
+        this.supportedProtocolTypes = supportedProtocolTypes;
+    }
+
+    public void process(Packet packet) {
+        Object bind = ioService.getSerializationService().toObject(packet);
+        TcpIpConnection connection = (TcpIpConnection) packet.getConn();
+        if (connection.setBinding()) {
```

Comment:
Looks like we may end up with the stuck connection if something would go wrong while sending the reply.

Reply:
https://github.com/hazelcast/hazelcast/issues/14497

Reply:
Indeed we may end up with a stuck connection in this case:
- client-side of connection sends bind request
- server-side of connection fails to send back reply
- client-side still attempts to `getOrConnect` a connection to the server, but the connection is already found in `connectionsInProgress` so no new connection is established and no additional bind request is sent -> no real progress can be achieved

For this scenario to manifest, the failure of server-side to send back reply must occur in such a way that is undetectable by the client-side of the connection (eg connection is still established). I am not sure if such a failure could occur in real world scenario. It is also worth considering that the existing behaviour of Hazelcast < 3.12 is the same (connection lingers in client-side `ConnectionManager#connectionsInProgress`, no progress can be made). In order to get this fixed, we should introduce some sort of timeout and check it when a connection is found in progress in `getOrConnect`. Then it would make sense to either retry the bind from the client-side (and restore the value of `connection.binding` on the server-side) or close the connection and let it be established again as needed. wdyt @taburet @tkountis ?

Reply:
I think we may live with it, if it occurs only in that rare specific case while a connection is established, but no data is coming from it. Most likely this indicates that something really bad is happened on the remote side and most probably that bad thing may be resolved only by performing a member restart, in this case we will lose the connection to it anyway.

Reply:
@vbekiaris I also agree with @taburet. 
Since this is a pre-existing issue, I believe we can live with it for now. I also like your proposal, but maybe we could target this for a later release.